File: go/binary_search/binary_search_count.go
package binary_search

// countInt should return number of values in a sorted array
func countInt(arr []int, val int) int {
	l, r := 0, len(arr)

	for l < r {
		m := (l + r) / 2

		if arr[m] < val {
			l = m + 1
		} else {
			r = m
		}
	}

	start := l

	l, r = 0, len(arr)

	for l < r {
		m := (l + r) / 2

		if arr[m] <= val {
			l = m + 1
		} else {
			r = m
		}
	}

	return r - start
}

File: java/src/binary_search/BinarySearchCount.java
package binary_search;

public class BinarySearchCount {
	// countInt should return number of values in a sorted array
	public static int countInt(int[] arr, int val) {
		int l = 0, r = arr.length;

		while (l < r) {
			int m = (l + r)/2;
			if (arr[m] >= val) {
				r = m;
			} else {
				l = m + 1;
			}
		}

		int l1 = l;
		r = arr.length;

		while (l < r) {
			int m = (l + r)/2;
			if (arr[m] > val) {
				r = m;
			} else {
				l = m + 1;
			}
		}

		return l - l1;
	}
}
